www/js/lib/uiUtil.js

```diff
@@ -51,15 +51,12 @@ define(rqDef, function(settingsStore) {
             // DEV: Note that webpMachine is single threaded and will reject an image if it is busy
             // However, the loadImagesJQuery() function in app.js is sequential (it waits for a callback
             // before processing another image) so we do not need to queue WebP images here
-            webpMachine.decode(content).then(function (uri) {
-                // DEV: WebpMachine.decode() returns a data: URI
-                // We callback before the node is set so that we don't incur slow DOM rewrites before processing more images
-                if (callback) callback(uri);
-                node.setAttribute(nodeAttribute, uri);
-            }).catch(function (err) {
-                console.error('There was an error decoding image in WebpMachine', err);
-                if (callback) callback();
-            });
+            var canvas = document.createElement("canvas");
+            webpMachine.decodeToCanvas(canvas, content);
```

Comment:
If there is a way to detect that the image has been decoded on the canvas, or loaded into it, we could Promisify this ourselves quite easily. I haven't tested anything, so only speculating at this stage.

Reply:
@mossroy I just tested refactoring this and treating `webpMachine.decodeToCanvas()` as a Promise. I tested it in IE11 (which does not support WebP), and it is working fine. It does appear to return a Promise, and looking at the source code, it uses an async function (here: https://github.com/chase-moskal/webp-hero/blob/master/source/webp-machine.ts#L43) which would normally be rendered by TypeScript as a Promise when bundled. Obviously the formatting is wonky because we're not removing the image nodes yet, but all the images appear to be there, and no error is thrown with the Promise.
![image](https://user-images.githubusercontent.com/4304337/153727572-8ee7c480-4c8a-4a53-840c-7ea9eb3c487e.png)

Reply:
Code I used:
```
var canvas = document.createElement("canvas");
webpMachine.decodeToCanvas(canvas, content).then(function () {
    // TODO copy all the img attributes to the new canvas, and remove the img node afterwards
    node.parentNode.appendChild(canvas);
    // TODO we don't handle images sequentially
    if (callback) callback();
});
```

Reply:
Alternative to inserting the canvas in the document is to use canvasElement.toBlob() if it isn't also blacklisted by the anti-fingerprinting in IceCat. See https://caniuse.com/mdn-api_htmlcanvaselement_toblob . It has to have an ms- prefix for IE11 but is otherwise supported.

Reply:
Cool! Thanks @Jaifroid 
I could not test but you're certainly right about the Promise. I got confused by the typescript transpilation to javascript.
Regarding the `toBlob` API, it's worth testing but there are high chances it will be "blocked" too

Reply:
@Jaifroid during my recent work on `webp-hero`, i had found that icecat appears to be blocking both `canvas.toDataURL` and also `canvas.toBlob()`, which is why i started inserting the whole canvas elements into the dom.

what i hadn't tried, was [canvas.mozGetAsFile(name, type)](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/mozGetAsFile) – it's a deprecated function, but it might be available in icecat, might be worth looking at.

Reply:
@chase-moskal Thank you! We can try that as this is only needed for old Mozilla browsers. I tried `canvas.toBlob()`, but as @mossroy suggested, it is (of course) blocked. In fact, any access to canvas properties appears to be blocked, which is a problem because we can't set the width/height, or at least I've so far failed to do so. I'm getting this kind of result (screenshot below), and if I try to set the width/height either with `canvas.width='12'0'` or` canvas.setAtrribute('width', '120')`, the images get blanked out.

![image](https://user-images.githubusercontent.com/4304337/153743415-9690b338-1cd7-4a01-8a40-9bc887dba0c7.png)

Reply:
It looks like `canvas.mozGetAsFile` is also locked down 😞. The resulting File objects can be converted to blobs, but they are empty. I guess this isn't surprising.

![image](https://user-images.githubusercontent.com/4304337/153744051-4d4b291a-86b5-49de-b20c-b9b62db3c177.png)